File: module_common/src/main/java/com/viomi/ovensocommon/rxbus/RxSchedulerUtil.java
package com.viomi.ovensocommon.rxbus;

import io.reactivex.FlowableTransformer;
import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;


/**
 * 统一管理观察者模式中线程调度（使用 RxJava2 过程中会频繁
 * 调用 subscribeOn() 和 observeOn()，通过 Transformer 结合
 * Observable.compose() 可实现复用）
 * <p>
 * Created by William on 2018/1/3.
 */
public final class RxSchedulerUtil {

    public static <T> ObservableTransformer<T, T> SchedulersTransformer1() {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> FlowableTransformer<T, T> SchedulersFlowableTransformer1() {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> ObservableTransformer<T, T> SchedulersTransformer2() {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io());
    }

    public static <T> FlowableTransformer<T, T> SchedulersFlowableTransformer2() {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io());
    }

    public static <T> ObservableTransformer<T, T> SchedulersTransformer3() {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation());
    }

    public static <T> FlowableTransformer<T, T> SchedulersFlowableTransformer3() {
        return upstream -> upstream.subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation());
    }

    public static <T> ObservableTransformer<T, T> SchedulersTransformer4() {
        return upstream -> upstream.subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> FlowableTransformer<T, T> SchedulersFlowableTransformer4() {
        return upstream -> upstream.subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> ObservableTransformer<T, T> SchedulersTransformer5() {
        return upstream -> upstream.subscribeOn(Schedulers.computation())
                .observeOn(Schedulers.io());
    }

    public static <T> FlowableTransformer<T, T> SchedulersFlowableTransformer5() {
        return upstream -> upstream.subscribeOn(Schedulers.computation())
                .observeOn(Schedulers.io());
    }

    public static <T> ObservableTransformer<T, T> SchedulersTransformer6() {
        return upstream -> upstream.subscribeOn(Schedulers.computation())
                .observeOn(Schedulers.computation());
    }

    public static <T> ObservableTransformer<T, T> SchedulersTransformer7() {
        return upstream -> upstream.subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: common_iot_device/src/main/java/com/viomi/iotdevice/common/util/RxSchedulerUtil.kt
package com.viomi.iotdevice.common.util

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.FlowableTransformer
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableTransformer
import io.reactivex.rxjava3.schedulers.Schedulers


/**
 * <pre>
 *     author : William
 *     e-mail : liangzhiwei@viomi.com.cn
 *     time   : 2018/1/3
 *     desc   : 统一管理观察者模式中线程调度（使用 RxJava2 过程中会频繁
 *              调用 subscribeOn() 和 observeOn()，通过 Transformer 结合
 *              Observable.compose() 可实现复用）
 *     version: 1.0
 * </pre>
 */
object RxSchedulerUtil {
    fun <T> schedulerTransformer1(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream: Observable<T> -> upstream.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()) }
    }

    fun <T> schedulerFlowableTransformer1(): FlowableTransformer<T, T> {
        return FlowableTransformer { upstream: Flowable<T> -> upstream.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()) }
    }

    fun <T> schedulerTransformer2(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream: Observable<T> -> upstream.subscribeOn(Schedulers.io()).observeOn(Schedulers.io()) }
    }

    fun <T> schedulerFlowableTransformer2(): FlowableTransformer<T, T> {
        return FlowableTransformer { upstream: Flowable<T> -> upstream.subscribeOn(Schedulers.io()).observeOn(Schedulers.io()) }
    }

    fun <T> schedulerTransformer3(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream: Observable<T> -> upstream.subscribeOn(Schedulers.io()).observeOn(Schedulers.computation()) }
    }

    fun <T> schedulerFlowableTransformer3(): FlowableTransformer<T, T> {
        return FlowableTransformer { upstream: Flowable<T> -> upstream.subscribeOn(Schedulers.io()).observeOn(Schedulers.computation()) }
    }

    fun <T> schedulerTransformer4(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream: Observable<T> -> upstream.subscribeOn(Schedulers.computation()).observeOn(AndroidSchedulers.mainThread()) }
    }

    fun <T> schedulerFlowableTransformer4(): FlowableTransformer<T, T> {
        return FlowableTransformer { upstream: Flowable<T> -> upstream.subscribeOn(Schedulers.computation()).observeOn(AndroidSchedulers.mainThread()) }
    }

    fun <T> schedulerTransformer5(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream: Observable<T> -> upstream.subscribeOn(Schedulers.computation()).observeOn(Schedulers.io()) }
    }

    fun <T> schedulerFlowableTransformer5(): FlowableTransformer<T, T> {
        return FlowableTransformer { upstream: Flowable<T> -> upstream.subscribeOn(Schedulers.computation()).observeOn(Schedulers.io()) }
    }

    fun <T> schedulerTransformer6(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream: Observable<T> -> upstream.subscribeOn(Schedulers.computation()).observeOn(Schedulers.computation()) }
    }

    fun <T> schedulerTransformer7(): ObservableTransformer<T, T> {
        return ObservableTransformer { upstream: Observable<T> ->
            upstream.subscribeOn(AndroidSchedulers.mainThread()).observeOn(AndroidSchedulers.mainThread())
        }
    }
}
